From 6e3da52c23bd27d02e27ff8e99408663fa4cf16a Mon Sep 17 00:00:00 2001
From: "han.liu" <liuhan907@163.com>
Date: Wed, 10 Mar 2021 17:18:24 +0800
Subject: [PATCH 1/2] plugin/forward: support except-file argument

---
 plugin/forward/setup.go | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/plugin/forward/setup.go b/plugin/forward/setup.go
index 7504e940..078e2cde 100644
--- a/plugin/forward/setup.go
+++ b/plugin/forward/setup.go
@@ -4,7 +4,10 @@ import (
 	"crypto/tls"
 	"errors"
 	"fmt"
+	"io/ioutil"
+	"path/filepath"
 	"strconv"
+	"strings"
 	"time"
 
 	"github.com/coredns/caddy"
@@ -154,6 +157,20 @@ func parseBlock(c *caddy.Controller, f *Forward) error {
 			ignore[i] = plugin.Host(ignore[i]).Normalize()
 		}
 		f.ignored = ignore
+	case "except-file":
+		args := c.RemainingArgs()
+		if len(args) != 1 {
+			return c.ArgErr()
+		}
+		b, err := ioutil.ReadFile(filepath.Clean(args[0]))
+		if err != nil {
+			return err
+		}
+		ignore := strings.Split(string(b), "\n")
+		for i := 0; i < len(ignore); i++ {
+			ignore[i] = plugin.Host(ignore[i]).Normalize()
+		}
+		f.ignored = ignore
 	case "max_fails":
 		if !c.NextArg() {
 			return c.ArgErr()
-- 
2.27.0.windows.1

